### PR TITLE
Add type hints

### LINKED ISSUE #93: provide type annotations
Maybe after removing support for older versions of Python.

Ref:
- https://docs.python.org/3/library/typing.html
- https://mypy.readthedocs.io/en/stable/generics.html#defining-sub-classes-of-generic-classes
- MagicStack/immutables#13

### PR DESCRIPTION
**Pull request by jab**
- Add type hints 🎉
- Run mypy in CI
- Add `bidict.MutableBidirectionalMapping`
- Drop support for Python 3.5.
- Remove `bidict.compat` now that it's no longer needed
- Additional minor improvements

Fixes #93.

### CODE REVIEW

**Comment by jab** [comment]
@cjgibson, @itaisteinherz, @martinthoma, @lordmauve, @gaborbernat, et al., would you like to try this branch and let me know if the type hints here are working well for you before I merge this?

**Comment by jab** [comment]
Doesn't have to block merging, but there are a few places where I used `# type: ignore` internally that may actually be worth type checking. Maybe. (In many of these places though I suspect appeasing mypy would not be worth it, e.g. `namedbidict` and `_invcls`.) Open to suggestions.

One thing I had trouble with was using `@typing.overload` with `__init__()` and `update()`. For example, mypy rejects:
```python
    @_t.overload
    def update(self, **kw: VT) -> None: ...
    @_t.overload
    def update(self, map: _t.Mapping[KT, VT], **kw: VT) -> None: ...
    @_t.overload
    def update(self, col: IterItems[KT, VT], **kw: VT) -> None: ...
    def update(self, *args: MapOrIterItems[KT, VT], **kw: VT) -> None:
        # (implementation here)
```
I think this may be a current limitation in mypy, so I left those without the overloads. If someone wants to suggest an alternative, I'll be happy to take another look.

**Comment by MartinThoma** [comment]
Sorry, I'm currently on vacation and have only limited time to check this thoroughly. But I appreciate the effort 🤗

File: .pylintrc [hunk 1]
```diff
@@ -1,17 +1,32 @@
...
-ignore = _version.py
-
-# Maximum number of parents for a class.
-# The default of 7 results in "too-many-ancestors" for all bidict classes.
-max-parents = 13
-
-# Maximum number of arguments for a function.
-# The default of 5 only leaves room for 4 args besides self for methods.
-max-args=6
+  protected-access,
```
Comment (gaborbernat) [inline .pylintrc:23]: Does this need to be in this PR? Seems unrelated a lot of them.

File: .travis.yml [hunk 2]
```diff
@@ -27,6 +27,10 @@ before_install:  # Ensure we have the required versions of Python and Pip.
   - 'echo "TRAVIS_PULL_REQUEST_SHA: $TRAVIS_PULL_REQUEST_SHA"'
   - 'echo "TRAVIS_COMMIT: $TRAVIS_COMMIT"'
   - 'git --no-pager log -n2'
+  - 'pwd'
+  - 'ls -la'
```
Comment (gaborbernat) [inline .travis.yml:31]: you probably want to remove these

File: bidict/__init__.py [hunk 3]
```diff
@@ -50,49 +50,43 @@
...
-)
+from ._dup import ON_DUP_DEFAULT, ON_DUP_RAISE, ON_DUP_DROP_OLD, RAISE, DROP_OLD, DROP_NEW, OnDup, OnDupAction
+from ._exc import BidictException, DuplicationError, KeyDuplicationError, ValueDuplicationError, KeyAndValueDuplicationError
 from ._util import inverted
 from .metadata import (
     __author__, __maintainer__, __copyright__, __email__, __credits__, __url__,
     __license__, __status__, __description__, __keywords__, __version__, __version_info__,
 )
 
+# Set __module__ of re-exported classes to 'bidict' for nicer class repr strings:
```
Comment (gaborbernat) [inline bidict/__init__.py:80]: This feels wrong, can you provide an example of why we need this?

File: bidict/_base.py [hunk 4]
```diff
@@ -26,30 +26,31 @@
...
+
+BT = _t.TypeVar('BT', bound='BidictBase')  # typevar for BidictBase.copy
 
 
-class BidictBase(BidirectionalMapping):
+class BidictBase(BidirectionalMapping[KT, VT]):
     """Base class implementing :class:`BidirectionalMapping`."""
 
-    __slots__ = ('_fwdm', '_invm', '_inv', '_invweak', '_hash', '__weakref__')
+    __slots__ = ['_fwdm', '_invm', '_inv', '_invweak', '_hash', '__weakref__']
```
Comment (gaborbernat) [inline bidict/_base.py:53]: unrelated?

File: tox.ini [hunk 5]
```diff
@@ -1,7 +1,7 @@
 [tox]
-envlist = py35, py36, py37, py38, pypy3
+envlist = py36, py37, py38, pypy3
 
 [testenv]
 commands =
-    pip install -e .[test]
+    pip install -e .[test] -U --upgrade-strategy=eager
```
Comment (gaborbernat) [inline tox.ini:6]: this tox configuration is wrong on many levels... prefer:

```
[tox]
envlist = 
    py38
    py37
    py36
    pypy3

isolated_build = true

[testenv]
setenv = 
  PIP_UPGRADE_STRATEGY=eager
extras = test
commands =
    python run_tests.py
```

And what's this ``run_test.py``? Couldn't you just do:

```
commands = 
   pytest tests
   sphinx-build -b doctest -d docs/_build/doctrees docs docs/_build/doctest
```
?

File: .pylintrc (hunk 1 above)
Comment (jab) [reply .pylintrc:23]: Thanks for taking a look at this PR!

Adding type hints to an entire codebase touches the entire codebase. In the process, I had to tune linter settings so I could spend less time fighting linters and be more productive. Specifically, linters fight with type hints and mypy in a couple ways: they lack good support composing line-based controls (e.g. `# noqa: E123; type: ignore; pylint: disable=foo`), and they flag idiomatic type hint style (e.g. for overloads).

In general, since I have had no one who is consistently available to review my changes to bidict (I am often the only one reviewing my PRs), and the code is all written by me alone on a voluntary basis on nights and weekends, it just hasn't been at all practical to optimize my PRs for the type of reviewer who prefers PRs to be split. I don't think I'm going to break this PR up into multiple PRs at this time. But it's helpful to know you prefer this in case you end up reviewing more of my PRs in the future.

File: .travis.yml (hunk 2 above)
Comment (jab) [reply .travis.yml:31]: No, I deliberately left these in because I'd like to keep an eye on what's getting cached by Travis's cacher.

File: bidict/_base.py (hunk 4 above)
Comment (jab) [reply bidict/_base.py:53]: No, not unrelated. If you read the rest of the changes in this file, you'll see this is for the https://bugs.python.org/issue41451 workaround below (removing the `'__weakref__'` slot) after this class definition.

File: tox.ini (hunk 5 above)
Comment (jab) [reply tox.ini:6]: run_tests.py allows for not duplicating the pytest and the sphinx-build commands across different call sites (such as this tox config and the Travis CI config).

I don't actually use tox much, but it's not clear what's preferable about this (e.g. in general I prefer explicit command line arguments over environment variables). In any case, thank you for the suggestions, I'll consider tweaking the tox config in a separate set of changes.

Thanks again for taking a look at this PR, but so far I haven't found your review very helpful. I am looking for reviewers to focus more on the type hints and in particular with the questions I called out in my comment above. Is that something you can help with?

Comment (gaborbernat) [reply tox.ini:6]: Why not call tox in Travis as all other projects do?

Comment (gaborbernat) [reply tox.ini:6]: Note that I'm author/maintainer of the tox project, so probably I'm more sensitive on this than other people would be. See a Github Actions variant of using tox https://github.com/pypa/virtualenv/blob/master/.github/workflows/check.yml#L22

Comment (jab) [reply tox.ini:6]: Thanks for the link!

File: bidict/__init__.py (hunk 3 above)
Comment (jab) [reply bidict/__init__.py:80]: Without this the private module gets exposed to the user:
```python
>>> from bidict import bidict
>>> bidict
<class 'bidict._bidict.bidict'>
```
Users should not be exposed to this implementation detail. The repr users see should match the module users are supposed to use. This change accomplishes that:
```python
>>> from bidict import bidict
>>> bidict
<class 'bidict.bidict'>
```
